### PR TITLE
Checking In YAML definition of existing Build pipeline

### PR DESCRIPTION
**Pull request by marstr**
This PR largely brings us to parity between the two following Azure DevOps builds:
- [devdiv.visualstudio.com](https://devdiv.visualstudio.com/DevDiv/_build?definitionId=10072)
- [dev.azure.com](https://dev.azure.com/azure-public/azcli/_build?definitionId=18&_a=summary)

The reasons to do this are:
1. Checking in our build-definition will allow us to track and test changes to our build defintion.
1. Allow us to consolidate all of our Build/Release automation into a single DevOps project, instead of being scattered around Jenkins, AppVeyor, DevOps, and Travis.
1. Diversify the medium that edge builds are available in. For instance, for each edge build we may be able to have:
    1. Pip packages
    1. Publish Docker images of varying Linux distributions to a container registry.
    1. MacOS Homebrew kegs.
    1. A Windows MSI installer

### CODE REVIEW

**Comment by marstr** [comment]
For the most part, all I did was literally take the automatically generated YAML definition of our existing DevOps build, and just check throw it in our repository. However, there were a handful things I had to change:
- The Homebrew Build phase was downloading a zip of the source code from GitHub. I streamlined this to just have a phase at the beginning of the job which creates both a zip file and a tarball.
- There were some secrets I had to scrub out.
- Azure DevOps generated some garbage that it couldn't even parse, and I had to troubleshoot that.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -0,0 +1,761 @@
...
+#Your build pipeline references an undefined variable named ‘cd $(dirname $0’. Create or edit the build pipeline for this YAML file, define the variable on the Variables tab. See https://go.microsoft.com/fwlink/?linkid=865972
+#Your build pipeline references an undefined variable named ‘cd $BUILD_SOURCESDIRECTORY/scripts/release/homebrew/docker; pwd’. Create or edit the build pipeline for this YAML file, define the variable on the Variables tab. See https://go.microsoft.com/fwlink/?linkid=865972
+#Your build pipeline references an undefined variable named ‘cd $BUILD_SOURCESDIRECTORY/scripts/release/homebrew/docker; pwd’. Create or edit the build pipeline for this YAML file, define the variable on the Variables tab. See https://go.microsoft.com/fwlink/?linkid=865972
+#Your build pipeline references an undefined variable named ‘cd $SYSTEM_ARTIFACTSDIRECTORY/pypi; pwd’. Create or edit the build pipeline for this YAML file, define the variable on the Variables tab. See https://go.microsoft.com/fwlink/?linkid=865972
```
Comment (tjprescott) [inline azure-pipelines.yml]: What's up with these comments?

File: azure-pipelines.yml [hunk 2]
```diff
@@ -0,0 +1,761 @@
...
+
+
+- phase: Phase_15
+  displayName: Test Python Wheels
+
+  dependsOn: Phase_1
+  condition: succeeded()
+  queue:
+    name: Hosted Ubuntu 1604
+#Your build pipeline references an undefined variable named ‘cd $SYSTEM_ARTIFACTSDIRECTORY/pypi; pwd’. Create or edit the build pipeline for this YAML file, define the variable on the Variables tab. See https://go.microsoft.com/fwlink/?linkid=865972
```
Comment (tjprescott) [inline azure-pipelines.yml]: This?

**Comment by tjprescott** [comment]
@williexu I don't think raising the verbosity of the script should be any problem.

File: azure-pipelines.yml [hunk 3]
```diff
@@ -0,0 +1,761 @@
...
+
+       docker load < $TAR_FILE
+       docker run $IMAGE_NAME /bin/bash -c "time az self-test && time az --version && sleep 5"
+    displayName: 'Bash Script'
+
+
+- phase: Phase_1
+  displayName: Build Python Wheels
+
+  dependsOn: Phase_13
```
Comment (tjprescott) [inline azure-pipelines.yml]: Phase_1 depends on Phase_13?  While these were probably autogenerated names, we should try to make them reflect reality. I think this file would already be confusing to most people. Things like this make it more confusing than it needs to be.

File: azure-pipelines.yml (hunk 1 above)
Comment (marstr) [reply azure-pipelines.yml]: A little bit unclear, seeing as they would suggest that the original definition shouldn't be working, or at very least is stashing intermediate progress in the wrong places. However, I've updated them to refer to variables that are defined by DevOps in d7ba697.

File: azure-pipelines.yml (hunk 3 above)
Comment (marstr) [reply azure-pipelines.yml]: Fixed in d7ba697

File: azure-pipelines.yml (hunk 1 above)
Comment (tjprescott) [reply azure-pipelines.yml]: Should the previous commit have failed the build then?  They were successful despite this warning.

Comment (marstr) [reply azure-pipelines.yml]: I don't _believe_ they should have failed. I suspect in each case, the empty string was being substituted for the expected value, and stashing the intermediate build products somewhere else proved unproblematic. Updating these values will drive consistency though.

**Pull request merged**